# DataMining Homework 3

In [6]:
import numpy as np
from scipy import stats

## Problem 1

Construct the matrix from data

In [3]:
x = [[1,1,1,1,1,1,1],[26.7,26.9,27.6,26.9,27.2,27.0,26.2],[1.14,1.41,1.13,1.37,1.14,1.26,1.28]]
x = np.matrix(x).T
x

matrix([[ 1.  , 26.7 ,  1.14],
        [ 1.  , 26.9 ,  1.41],
        [ 1.  , 27.6 ,  1.13],
        [ 1.  , 26.9 ,  1.37],
        [ 1.  , 27.2 ,  1.14],
        [ 1.  , 27.  ,  1.26],
        [ 1.  , 26.2 ,  1.28]])

In [4]:
y = np.matrix([87.8, 87, 88, 87.1, 87.9, 87, 88])
y = y.T
y

matrix([[87.8],
        [87. ],
        [88. ],
        [87.1],
        [87.9],
        [87. ],
        [88. ]])

Using the formula \\( \hat\beta = (X^{T}X)^{-1}X^{T}y\\) <br>
Calculate estimate beta 

In [5]:
#Calculating Beta
beta = np.linalg.inv(x.T*x) * x.T * y
beta

matrix([[102.5419763 ],
        [ -0.38347545],
        [ -3.7467023 ]])

In [6]:
x * beta

matrix([[88.03194106],
        [86.94363635],
        [87.72428018],
        [87.09350445],
        [87.84020334],
        [87.46729415],
        [87.69914047]])

Calculating estimate variance

In [7]:
residual = y - x * beta
residual

matrix([[-0.23194106],
        [ 0.05636365],
        [ 0.27571982],
        [ 0.00649555],
        [ 0.05979666],
        [-0.46729415],
        [ 0.30085953]])

In [8]:
SSE = residual.T * residual
SSE

matrix([[0.44549305]])

In [9]:
df = 7-3
estimate_var = SSE / df
estimate_var

matrix([[0.11137326]])

In [16]:
C = np.linalg.inv(x.T * x)
C

matrix([[888.41039039, -30.24486693, -59.18846697],
        [-30.24486693,   1.05299588,   1.51481626],
        [-59.18846697,   1.51481626,  14.75101981]])

Calculating standard error for each regression coefficients

In [20]:
se_beta0 = estimate_var * C[0,0]
se_beta0 = np.sqrt(se_beta0)
se_beta0

matrix([[9.94711839]])

In [21]:
se_beta1 = estimate_var * C[1,1]
se_beta1 = np.sqrt(se_beta1)
se_beta1

matrix([[0.34245523]])

In [22]:
se_beta2 = estimate_var * C[2,2]
se_beta2 = np.sqrt(se_beta2)
se_beta2

matrix([[1.2817446]])

Perform a t-test for each regression coefficient

In [23]:
from scipy import stats

In [24]:
beta_1 = beta[1,0]
beta_2 = beta[2,0]

In [25]:
t_1 = beta_1 / se_beta1
t_2 = beta_2 / se_beta2

In [26]:
t_1

matrix([[-1.11978272]])

In [27]:
t_2

matrix([[-2.92312705]])

In [28]:
stats.t.interval(0.95,df=df)

(-2.7764451051977987, 2.7764451051977987)

In [29]:
stats.t.ppf(0.975, df=df)

2.7764451051977987

Here t_1 < 2.77644 we can not reject the null hypothesis \\(H_0 : \beta_1 = 0 \\)

In case of \\(\beta_2 , \ \   t_2 < -2.77644\\), we can reject the null hypothesis which states, \\(H_0 : \beta_2 = 0 \\)

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/ad6b035acd8f46d2971dbbfaa09b4b35/image.png">

In [30]:
SSE

matrix([[0.44549305]])

In [31]:
SST = (y - y.mean()).T * (y - y.mean())
SST

matrix([[1.39714286]])

In [32]:
SSR = SST - SSE
SSR

matrix([[0.9516498]])

In [33]:
# 2 : number of predictor varible
MSR = SSR / 2
MSR

matrix([[0.4758249]])

In [34]:
# 4 : 7 - (2+1)
MSE = SSE / 4
MSE

matrix([[0.11137326]])

In [35]:
F = MSR / MSE
F

matrix([[4.27234408]])

ppf : Percent point function (inverse of cdf — percentiles).

In [36]:
stats.f.ppf(0.95, 2, 4)

6.944271909999155

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/cf743a41b3c38c2804209d389950ce16/image.png">

In [37]:
R_square = SSR/SST
R_square

matrix([[0.68113994]])

\\({R}^{2}= 0.68113994 \\)

Since F=4.2723 < 6.94427 we can not reject the null_hypothesis

## Problem 2

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/73dd1061dee00413f9b94c1156ef5bd6/image.png">

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/474bd92600480b7e22ed9f90578a419c/image.png">

### (A) Done

### (B)  \\({R}^{2}= \frac{SSR}{SST} = \frac{233.6}{835.97} = 0.27943586 \\)

In [32]:
233.6/835.97

0.2794358649233824

### (C)	Test for the regression model at the significance level of 1%.

In [33]:
stats.f.ppf(0.99, 4, 188)

3.4205017271207225

Since 3.4205 < 18.2272 = F-Value, we can reject the null hypothesis \\(H_0 \\)

Where \\(H_0 : \beta_1 = \beta_2 = \beta_3 = \beta_4 = 0\\)

### (D)	Which regression coefficient is significant at the significance level of 1%?

We can select the parameter whose p-value is less than 0.01

Answer is A and D

### (E)	Obtain the confidence interval of the regression coefficient of A at the significance level of 5%.

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/41ee87f3aef815aa68f55ba1081e5825/image.png">

In [14]:
t = stats.t.ppf(0.975,188)
t

1.9726626923781652

In [15]:
print('The 95% confidence interval for coefficient of A')
print('(', -0.014 - t * 0.00476, ',', -0.014 + t * 0.00476, ')')

The 95% confidence interval for coefficient of A
( -0.023389874415720067 , -0.004610125584279933 )


### (F) Estimate the response value at A=546, B=765, C=730, D=12

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/194a8e75beffde5b4701b01a179debcf/image.png">

If we know the data set information, we can get the \\(100(1-\alpha) \\) confidence interval for the response value using above formula, but in here, we don't have that information so we could not calculate \\(V=(X^{T}X)^{-1} \\)

In [40]:
coef = np.array([78.941, -0.014, 0.012, -0.020, 0.584])

In [41]:
data_in = np.array([1, 546, 765, 730, 12])

In [47]:
estimate_mean = np.dot(coef, data_in)
estimate_mean

72.884999999999991

## Problem 3

### (a) calculating p_value

In [62]:
p_value = 1 - stats.f.cdf(298,3,96)
p_value

1.1102230246251565e-16

### (b)	Test for the regression coefficients at the significance level of 5%.

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/13dbe319aa0da6cd19448f59efcb07bf/image.png">

Coefficient for \\(C\\)

In [65]:
p_value = 1 - stats.t.cdf(4.8,96)
p_value

2.9117612861950803e-06

- Since p-value is further less than 0.025 we can reject the null hypothesis \\(H_0 : \beta_1 = 0 \\)

Coefficient for \\(Mn \\)

In [66]:
p_value = 1 - stats.t.cdf(1.6,96)
p_value

0.056442284972185952

- Since p-value is larger than 0.025 (two-sided) we can not reject the null hypothesis \\(H_0 : \beta_2 = 0 \\)

Coefficient for \\(CT \\)

In [10]:
p_value = stats.t.cdf(-3.2,96)
p_value

0.0009315631847482069

- Since p-value is less than 0.025 we can reject the null hypothesis \\(H_0 : \beta_3 = 0 \\)

### (C) Obtain the 95% confidence interval of the regression coefficient of CT.

From the equation \\(t_{value}= \frac{\hat\beta_j - 0}{SE(\hat\beta_j)}, \ \ SE(\hat\beta_3) = 0.0046875 \\)

In [13]:
t = stats.t.ppf(0.975,96)
t

1.984984311431769

In [41]:
print('The 95% confidence interval for coefficient of CT')
print('(', -0.014 - t * 0.0046875, ',', -0.014 + t * 0.0046875, ')')

The 95% confidence interval for coefficient of CT
( -0.02324685637052265 , -0.004753143629477352 )


### (D) Assuming that the regression coefficient of C is β_1, test the following at the significance level 5%.

\\(H_0 : \beta_1 = 80 \\) <br>
\\(H_1 : \beta_1 \not= 80 \\)

From the equation \\(t_{value}= \frac{\hat\beta_j - 0}{SE(\hat\beta_j)}, \ \ SE(\hat\beta_1) = 17.5417 \\)

New t value for this hypothesis is \\(t = \frac{\hat\beta_1 - 80}{SE(\hat\beta_1)} = 0.2394\\)

In [69]:
p_value = 1 - stats.t.cdf(0.2394,96)
p_value

0.40565290504997686

- Since p-value is larger than 0.025 we can not reject the null hypythesis \\(H_0 : \beta_1 = 80 \\)

## Problem 4

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/4aed82ec00f3aff8481acf5b7fe87db5/image.png">

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/6bf18a74ccbdbf7208aafde7c667ac8f/image.png">

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/a1065e65a9a712454133c071e21aa2c4/image.png">

Put all informations to calculating F-value

In [71]:
F = ((315.44 - 307.08) / 2) / (307.08/(92))
F

1.2523121010811535

Perform F-Test

In [74]:
p_value = 1 - stats.f.cdf(F,2,92)
p_value

0.29066940525261187

- Since p-value is larger than 0.05 we cannot reject the null hypothesis \\(H_0 : \beta_3 = \beta_4 = 0 \\) 

## Problem 5

<img src = "https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/c4db295c65e0368a99974b67d6f4f102/image.png">

### (b)	Determine whether the regression coefficient of cooling water temperature is 1.0 at the significance level of 5%

Let \\(\beta_2 \\) be the coefficient of colling water tempaerature. Then we can perform following hypothesis test.

\\(H_0 : \beta_2 = 1 \\) <br>
\\(H_1 : \beta_2 \not= 1 \\)

In [38]:
1.2953/3.52

0.3679829545454545

From the equation \\(t_{value}= \frac{\hat\beta_2 - 0}{SE(\hat\beta_2)}, \ \ SE(\hat\beta_2) = 0.368 \\)

In [39]:
0.2953/ 0.368

0.8024456521739131

New t value for this hypothesis is \\(t = \frac{\hat\beta_2 - 1}{SE(\hat\beta_2)} = 0.8024\\)

In [8]:
stats.t.interval(0.95,17)

(-2.1098155778331806, 2.1098155778331806)

Since, new \\(t=0.8024\\) is contained in an significance 5% interval, we can not reject the hypothesis \\(H_0 : \beta_2 = 1 \\)

### (C) ANOVA table

<img src = 'https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/9b88c5095c2a0753dcdfd57735f3bf1c/image.png'>

In [3]:
p_value = 1- stats.f.cdf(60.225,3,17)
p_value

2.8936110174626606e-09

<img src = 'https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/7a76cd9e5c2f22f632b853528ad0666a/image.png'>

## Problem 6

<img src = 'https://trello-attachments.s3.amazonaws.com/5b29ec749cfb0d90ada47d03/5beedd5582101b1b17d77423/c212152f380865cc167f4e947e9f1b48/image.png'>

### (a)	Find eigenvalues and eigenvectors for the above matrix.

In [43]:
import numpy as np

In [44]:
mat = np.matrix([24.8, -7.3, 15.1, -7.3, 11.2, 4.6, 15.1, 4.6, 39.4])
mat = mat.reshape(3,3)
mat

matrix([[24.8, -7.3, 15.1],
        [-7.3, 11.2,  4.6],
        [15.1,  4.6, 39.4]])

In [45]:
eig_val, eig_vec = np.linalg.eig(mat)

In [46]:
for i in range(3):
    print('Eigen Value ' + str(i+1) + ' : ', eig_val[i])
    print('Corresponding Eigen Vector : ', eig_vec[:,i].reshape(3))

Eigen Value 1 :  48.872009252072814
Corresponding Eigen Vector :  [[-5.31281578e-01 -4.97510157e-04 -8.47195159e-01]]
Eigen Value 2 :  22.135848762750303
Corresponding Eigen Vector :  [[ 0.66526499 -0.61941643 -0.41682824]]
Eigen Value 3 :  4.392141985176856
Corresponding Eigen Vector :  [[ 0.52455922  0.78506245 -0.32941551]]


### (b)	How many principal components should be used to account for about 90% of the total variance?

 If we take two principal components, \\(\frac{\lambda_1 + \lambda_2}{\sum\lambda}  = 0.9417 > 90\% \\) 

In [39]:
(eig_val[0] + eig_val[1]) / eig_val.sum()

0.9417487800374421

In [40]:
stats.f.ppf(0.95,3,96)

2.6993925975521815

### (c)	Compute the correlation coefficient matrix and find corresponding eigenvalues and eigenvectors.

\\( corr(X) = (diag(\sum))^{-\frac{1}{2}}\sum(diag(\sum))^{-\frac{1}{2}} \\)

In [70]:
norm = np.diag(np.diag(mat) ** -0.5)
norm = np.matrix(norm)
norm

matrix([[0.20080483, 0.        , 0.        ],
        [0.        , 0.29880715, 0.        ],
        [0.        , 0.        , 0.15931325]])

In [71]:
corr_mat = norm * mat * norm
corr_mat

matrix([[ 1.        , -0.43801402,  0.48306213],
        [-0.43801402,  1.        ,  0.21897811],
        [ 0.48306213,  0.21897811,  1.        ]])

In [72]:
eig_val, eig_vec = np.linalg.eig(corr_mat)

In [74]:
for i in range(3):
    print('Eigen Value ' + str(i+1) + ' : ', eig_val[i])
    print('Corresponding Eigen Vector : ', eig_vec[:,i].reshape(3))

Eigen Value 1 :  0.22964485457507644
Corresponding Eigen Vector :  [[-0.6459921  -0.52485814  0.55427262]]
Eigen Value 2 :  1.5527228830503557
Corresponding Eigen Vector :  [[-0.76218062  0.40341776 -0.50629519]]
Eigen Value 3 :  1.2176322623745688
Corresponding Eigen Vector :  [[-0.04212973  0.74951854  0.66064139]]


## Problem 7

### (a)	Estimate the coefficients in PCR with t1 as an independent variable and determine the coefficient of determination for this model

In [102]:
t_1 = np.matrix([72.0693, 49.3127, -14.2052, 70.0119, -14.3851, -6.133, -27.995, -11.6537, 
       25.1774, -24.4915, -37.2383, -39.1275, -32.7934, -41.8799, 33.3314])
t_1 = t_1.T

In [91]:
t_2 = np.matrix([29.0359, 28.9586, 18.1284, 35.6215, 1.6942, -25.8141, -0.0378, 
       -16.9481, -61.1165, 15.5325, 12.2785, 11.6816, 12.9149, 10.2159, -72.1454])
t_2 = t_2.T

In [103]:
# constant term
t_0 = np.matrix(np.ones(len(t_1))).T

In [90]:
y = np.matrix([163, 74, -135, 126, -84, 103, -116, 29, 397, -174, -211, -218, -200, -223, 469]).T

In [138]:
x = np.hstack((t_0,t_1))
x

matrix([[  1.    ,  72.0693],
        [  1.    ,  49.3127],
        [  1.    , -14.2052],
        [  1.    ,  70.0119],
        [  1.    , -14.3851],
        [  1.    ,  -6.133 ],
        [  1.    , -27.995 ],
        [  1.    , -11.6537],
        [  1.    ,  25.1774],
        [  1.    , -24.4915],
        [  1.    , -37.2383],
        [  1.    , -39.1275],
        [  1.    , -32.7934],
        [  1.    , -41.8799],
        [  1.    ,  33.3314]])

Using the formula \\( \hat\beta = (X^{T}X)^{-1}X^{T}y\\) <br>
Calculate estimate beta 

In [139]:
beta_hat = np.linalg.inv(x.T * x) * x.T * y
beta_hat

matrix([[-2.78360339e-05],
        [ 4.17540508e+00]])

In [140]:
residual = y - x * beta_hat
residual

matrix([[-137.91849349],
        [-131.90047025],
        [ -75.68750792],
        [-166.32801508],
        [ -23.93635255],
        [ 128.60778719],
        [   0.89049305],
        [  77.65894602],
        [ 291.87418398],
        [ -71.73803865],
        [ -55.51498518],
        [ -54.6268099 ],
        [ -63.07424322],
        [ -48.13442496],
        [ 329.82793096]])

In [141]:
SST = (y - y.mean()).T * (y - y.mean())
SST

matrix([[687728.]])

In [142]:
SSE = residual.T * residual
SSE

matrix([[304442.10358948]])

In [143]:
r_square = 1 - (SSE / SST)
r_square

matrix([[0.55732193]])

\\( R^{2} = 0.5573 \\)

### (b)	Estimate the coefficients in the PCR with t1 and t2 as independent variables and determine the coefficient of determination for this model.

In [144]:
x = np.hstack((t_0,t_1,t_2))
x

matrix([[ 1.00000e+00,  7.20693e+01,  2.90359e+01],
        [ 1.00000e+00,  4.93127e+01,  2.89586e+01],
        [ 1.00000e+00, -1.42052e+01,  1.81284e+01],
        [ 1.00000e+00,  7.00119e+01,  3.56215e+01],
        [ 1.00000e+00, -1.43851e+01,  1.69420e+00],
        [ 1.00000e+00, -6.13300e+00, -2.58141e+01],
        [ 1.00000e+00, -2.79950e+01, -3.78000e-02],
        [ 1.00000e+00, -1.16537e+01, -1.69481e+01],
        [ 1.00000e+00,  2.51774e+01, -6.11165e+01],
        [ 1.00000e+00, -2.44915e+01,  1.55325e+01],
        [ 1.00000e+00, -3.72383e+01,  1.22785e+01],
        [ 1.00000e+00, -3.91275e+01,  1.16816e+01],
        [ 1.00000e+00, -3.27934e+01,  1.29149e+01],
        [ 1.00000e+00, -4.18799e+01,  1.02159e+01],
        [ 1.00000e+00,  3.33314e+01, -7.21454e+01]])

Using the formula \\( \hat\beta = (X^{T}X)^{-1}X^{T}y\\) <br>
Calculate estimate beta 

In [145]:
beta_hat = np.linalg.inv(x.T * x) * x.T * y
beta_hat

matrix([[ 3.25266621e-06],
        [ 4.17540623e+00],
        [-4.66330616e+00]])

In [146]:
residual = y - x * beta_hat
residual

matrix([[ -2.51531608],
        [  3.14225973],
        [  8.85075665],
        [ -0.21416628],
        [-16.03579383],
        [  8.22871166],
        [  0.71422114],
        [ -1.37525078],
        [  6.86917319],
        [  0.69476128],
        [  1.74343116],
        [ -0.15191884],
        [ -2.84810395],
        [ -0.49453858],
        [ -6.60822647]])

In [147]:
SST = (y - y.mean()).T * (y - y.mean())
SST

matrix([[687728.]])

In [148]:
SSE = residual.T * residual
SSE

matrix([[524.5979642]])

In [149]:
r_square = 1 - (SSE / SST)
r_square

matrix([[0.9992372]])

\\( R^{2} = 0.9992 \\)

## Problem 8

### (c)

\\( p = \frac{exp(-9.72357 + 0.782704X1 + 0.535184X6 + 0.937970X7)}{1+exp(-9.72357 + 0.782704X1 + 0.535184X6 + 0.937970X7)} \\)

In [4]:
d = np.exp(-9.72357 + 0.782704 * 5 + 0.535184 * 10 + 0.937970 * 4)

In [5]:
d/(d+1)

0.9642110149521727

### (d)

probit(p) = -9.72357 + 0.782704 * 5 + 0.535184 * 10 + 0.937970 * 4

p = cdf of normal distribution at point (-9.72357 + 0.782704 * 5 + 0.535184 * 10 + 0.937970 * 4)

In [13]:
p = stats.norm.cdf(-9.72357 + 0.782704 * 5 + 0.535184 * 10 + 0.937970 * 4)
p

0.999505557387677

In [14]:
p = stats.norm.cdf(-9.72357 + 0.782704 * 9 + 0.535184 * 3 + 0.937970 * 5)
p

0.9998505018165776

## Problem 9

###  \\( P_i = \frac{exp(\beta_0 + \beta_1 x)}{1+exp(\beta_0 + \beta_1 x)} \\)

###  \\( 1- P_i = \frac{1}{1+exp(\beta_0 + \beta_1 x)}\\)

###  \\( \frac{P_i}{1-P_i} = exp(\beta_0 + \beta_1 x) \\)

\\( logit(p) = ln( \frac{P_i}{1-P_i} ) = \beta_0 + \beta_1x \\)

## Problem 12

Data A : 6,6,8 <br>
Data B : 1,5,14

In [17]:
gini_A = 1 - (6/20) ** 2 - (6/20) ** 2 - (8/20) ** 2
gini_A

0.66

In [19]:
gini_B = 1 - (1/20) ** 2 - (5/20) ** 2 - (14/20) ** 2
gini_B

0.4450000000000001

In [20]:
entropy_A = -((6/20)*np.log(6/20)+(6/20)*np.log(6/20)+(8/20)*np.log(8/20))
entropy_A

1.0888999753452238

In [21]:
entropy_B = -((1/20)*np.log(1/20)+(5/20)*np.log(5/20)+(14/20)*np.log(14/20))
entropy_B

0.7460326647147849

Since both Gini and Entropy is larger in the Data A, we can conclude that Data A's impurity is higher.

In [22]:
1-(1/6) **2 - (5/6) ** 2

0.2777777777777777

In [25]:
0.5 - 0.1667

0.33330000000000004

In [26]:
1-(7/12)**2-(5/12)**2

0.48611111111111094

## Problem 11

In [27]:
s_1 = np.matrix([[600, 238.8],[238.8, 225]])
s_2 = np.matrix([[500,176.8],[176.8, 250]])

In [35]:
s_w = (s_1 + s_2) / 250

In [37]:
np.linalg.inv(s_w) * np.matrix([[-4],[-4]])

matrix([[-0.16982266],
        [-1.95667727]])

In [27]:
import pandas as pd
import numpy as np

In [28]:
a = pd.Series([0.25,0.25,0.25,0.25])
a

0    0.25
1    0.25
2    0.25
3    0.25
dtype: float64

In [29]:
b = pd.Series([1,-0.5,1,-0.5])
b

0    1.0
1   -0.5
2    1.0
3   -0.5
dtype: float64

In [30]:
def get_CAGR(data):
    data = np.array(data) + 1
    return data.prod() ** (1/len(data))

In [31]:
get_CAGR(a)

1.25

In [32]:
get_CAGR(b)

1.0